py/mpconfig.h

```diff
@@ -597,6 +597,14 @@
 #define MICROPY_READER_VFS (0)
 #endif
 
+#ifndef MICROPY_READER_VFS_BUFFER_SIZE
```

Comment:
Increasing the buffer size should also improve import speed from the filesystem.  Would be good to measure that.  In that case we'd want to have a large-ish buffer even when not using mpremote.

So, I'm not sure an ioctl would be that useful?  Just have a fixed size.

Reply:
I also thought about the ROM/RAM implications of that setting, though I'm sure other features end up taking more/less ram in line with the rom level setting ;-)

ioctl isn't a bad idea, though it does add overhead. Also, do we use `mp_vfs_blockdev_ioctl` as a "global" vfs ioctl that would set a default buffer size for the filesystem in question, or the `IOBase.ioctl` and have to set it on a per-file basis?
blockdev doesn't quite feel right, it's very abstracted from the "file" level that this buffer is allocated for. But having to set the buffer size for each file seems quite annoying.

Reply:
My intention was it's a file ioctl (i.e. the vfs file that the reader is currently opening). The filesystem can then chose to implement that particular ioctl (or not). The fact that it's per-file doesn't matter, e.g. mpremote's `RemoteFile` can always return a constant value to that ioctl.

(There is no block device involved here)

Reply:
> Increasing the buffer size should also improve import speed from the filesystem. Would be good to measure that

I'm assuming this will be quite different on each platform, depending on flash speeds and interfaces. I reckon it won't make much different on stm's with spiflash as that generally has a buffer already, but as a first test with the esp32-s2, loading the same hell.py as earlier but from internal flash:
without change: `321ms` import
with (24+180): `233ms` import

Reply:
I also did some tests importing from the local filesystem, on stm32 (internal flash, PYBv1.0) and rp2.  In both cases increasing the reader buffer size has a very small improvement on import speed about 2-3% only.

Reply:
> My intention was it's a file ioctl (i.e. the vfs file that the reader is currently opening).

The issue is there's not really any file to execute the ioctl on.  At the Python level you call `import foo` and that constructs a VFS reader (allocating the read buffer) and then calls `mp_vfs_open()`, which ends up in the mpremote-mount code.  The latter has no idea about the VFS reader, nor any way to reference it.

Reply:
I started to get a feel for how this could work, `mp_reader_new_file` calls `mp_vfs_open` after which it could immediately call `bufsize = file->ioctl(GET_BUFFER_SIZE)` then `reader->buf = malloc(bufsize)`.

However it still requires the "file type" class ioctl to specify a suitable buffer size; the python impl remotefile in mpremote  doesn't know how big the buffer could be unless it queries the chip first to get an idea of free ram.

Reply:
@andrewleech Yes that is pretty much exactly the idea I'm suggesting. Except you don't need the double malloc, because you can use m_new_obj_var to allocate the mp_reader_vfs_t. (You just need to create the file first now).

> ; the python impl remotefile in mpremote doesn't know how big the buffer

It does -- it's the bigger value that we're proposing here. But it could also be a flag to mpremote which the user could set even larger if they want a small extra speed boost (or they know their board has lots of RAM).

(There's a mildly annoying detail which is that we now need to steal a byte or two to store the length of the buffer in the mp_reader_vfs_t struct).

Reply:
> It does -- it's the bigger value that we're proposing here

However if you're using a small device this larger value might not be appropriate! 

Also I did get ~30% speed up of import from flash for larger py file on some boards, which we'd miss out on if the internal file types didn't use the larger value.

Reply:
> However if you're using a small device this larger value might not be appropriate!

If you're using a small device, mpremote-mount probably doesn't work because it already requires a fair bit of RAM to load the VFS hook code.  If you have the RAM for mpremote-mount, you most likely want to spend a little bit more in the buffer to get acceptable speed.

I think @jimmo's idea with the ioctl is the way to go.  Because the buffer size is then dependent on the filesystem properties, not the board properties, which makes more sense (it's the filesystem that's slow, not the board).  Also, a user of mpremote could increase this buffer size via an mpremout-mount command-line flag, in case they want the mount as fast as possible and have the RAM to spare.

If/when we have a memory-map FS, that won't need any buffer, and so it doesn't need to implement the ioctl, and doesn't need to waste RAM on a buffer.

If some boards show increased speed of filesystem imports with a bigger buffer, then we can either increase the default buffer size on those boards, or have them implement the ioctl for FAT and littlefs files when they sit on top of a block device (because it's the underlying block device that is slow).